### PR TITLE
kubeadm UX improvements for the v1.5 stable release

### PR DESCRIPTION
**Pull request by luxas**
This PR targets the next stable kubeadm release.

It's work in progress, but please comment on it and review, since there are many changes.

I tried to group the commits logically, so you can review them separately.

Q: Why this large PR? Why not many small?
A: Because of the Submit Queue and the time it takes.

PTAL @kubernetes/sig-cluster-lifecycle

_Edit:_ This work was splitted up in three PRs in total

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37568)

### CODE REVIEW

**Comment by pires** [comment]
@luxas while I understand your frustration, the process is in place for everyone. Even if I'm not assigned to this, I wanted to review but in the end it a hard task. If at least we had unit/e2e tests..

**Comment by luxas** [comment]
@pires I'm not frustrated at all.
I just follow the pattern we have discussed in `kubernetes-dev`, that the fewer kubeadm PRs we have to merge this week, the better, given this rush with PRs we have right now.

This was a decision, but I definitely don't think this is too overwhelming to review, we reviewed ~5000 LOC in the initial, and now I've grouped the changes in some commits.

And in comparision: https://github.com/kubernetes/kubernetes/pull/36263 is twice the size of this one.

Also, this code is battle-tested e2e-wise (manually), I've used it for spinning up a lot of DigitalOcean clusters last week.

I will continue to work on it and rebase upon all changes I'm merging this week, and then get this up for final review and merge. But please choose a commit and start looking at it.

**Comment by pires** [comment]
> And in comparision: #36263 is twice the size of this one.

I won't measure PR complexity with LOC. The PR you linked is purely unit-testing, this is not. It touches a lot of the different pieces of code and so, to me, it's complicated to review properly.

Anyway, this is just my two cents. If you guys decided to do it, by all means! 👍

**Comment by luxas** [comment]
Read kubernetes-dev on Slack :), but again, commits are reviewable separately

**Comment by errordeveloper** [comment]
I agree with @pires.

**Comment by marun** [comment]
I don't think this PR is reviewable in its current form.  Even if it were acceptable to lump everything together - and I don't think it is - it is essential to ensure that the message associated with each commit clearly documents the intent of the code changes.  Without that coherency, a reviewer will have a difficult time providing useful feedback for commits like 'wip' and 'a lot of changes'.

**Comment by luxas** [comment]
Rebased and updated, this is still WIP, but the 7th, 8th and maybe 9th commit can now be reviewed.
The 6 first commits are from #37835 and #37831, so don't be scared of the size.

**Comment by luxas** [comment]
As soon as #37831 and #37835 merge, I'm gonna rebase/update this PR to be mergeable (probably tomorrow), but here is a sneak peek what it now looks like:

Old `kubeadm init` (terrible!):
```
Running pre-flight checks
I1206 14:53:12.397845   15515 validators.go:50] Validating os...
I1206 14:53:12.398782   15515 validators.go:50] Validating kernel...
I1206 14:53:12.399455   15515 kernel_validator.go:77] Validating kernel version
I1206 14:53:12.399539   15515 kernel_validator.go:92] Validating kernel config
I1206 14:53:12.409318   15515 validators.go:50] Validating cgroups...
I1206 14:53:12.409391   15515 validators.go:50] Validating docker...
Using Kubernetes version: v1.4.6
<master/tokens> generated token: "579bd7.3d2b5a1cd24b6964"
<master/pki> generated Certificate Authority key and certificate:
Issuer: CN=kubernetes | Subject: CN=kubernetes | CA: true
Not before: 2016-12-06 12:53:13 +0000 UTC Not After: 2026-12-04 12:53:13 +0000 UTC
Public: /etc/kubernetes/pki/ca-pub.pem
Private: /etc/kubernetes/pki/ca-key.pem
Cert: /etc/kubernetes/pki/ca.pem
<master/pki> generated API Server key and certificate:
Issuer: CN=kubernetes | Subject: CN=kube-apiserver | CA: false
Not before: 2016-12-06 12:53:13 +0000 UTC Not After: 2017-12-06 12:53:14 +0000 UTC
Alternate Names: [192.168.1.115 10.96.0.1 kubernetes kubernetes.default kubernetes.default.svc kubernetes.default.svc.cluster.local]
Public: /etc/kubernetes/pki/apiserver-pub.pem
Private: /etc/kubernetes/pki/apiserver-key.pem
Cert: /etc/kubernetes/pki/apiserver.pem
<master/pki> generated Service Account Signing keys:
Public: /etc/kubernetes/pki/sa-pub.pem
Private: /etc/kubernetes/pki/sa-key.pem
<master/pki> created keys and certificates in "/etc/kubernetes/pki"
<util/kubeconfig> created "/etc/kubernetes/kubelet.conf"
<util/kubeconfig> created "/etc/kubernetes/admin.conf"
<master/apiclient> created API client configuration
<master/apiclient> created API client, waiting for the control plane to become ready
<master/apiclient> all control plane components are healthy after 71.296436 seconds
<master/apiclient> waiting for at least one node to register and become ready
<master/apiclient> first node is ready after 1.502372 seconds
<master/apiclient> attempting a test deployment
<master/apiclient> test deployment succeeded
<master/apiclient> failed to delete test deployment [no kind "DeleteOptions" is registered for version "kubeadm.k8s.io/v1alpha1"] (will ignore)<master/discovery> created essential addon: kube-discovery, waiting for it to become ready
<master/discovery> kube-discovery is ready after 3.002425 seconds
<master/addons> created essential addon: kube-proxy
<master/addons> created essential addon: kube-dns

Kubernetes master initialised successfully!

You can now join any number of machines by running the following on each node:

kubeadm join --token=579bd7.3d2b5a1cd24b6964 192.168.1.115
```

New `kubeadm init`:
```
[kubeadm] Bear in mind that kubeadm is in alpha, do not use it in production clusters.
[preflight] Running pre-flight checks...
[preflight] Starting the kubelet service by running "systemctl start kubelet"
[init] Using Kubernetes version: v1.4.6
[tokens] Generated token: "e69295.cc499553b867df2c"
[certificates] Generated Certificate Authority key and certificate.
[certificates] Generated API Server key and certificate
[certificates] Generated Service Account signing keys
[certificates] Created keys and certificates in "/etc/kubernetes/pki"
[kubeconfig] Wrote KubeConfig file to disk: "/etc/kubernetes/kubelet.conf"
[kubeconfig] Wrote KubeConfig file to disk: "/etc/kubernetes/admin.conf"
[apiclient] Created API client, waiting for the control plane to become ready
[apiclient] All control plane components are healthy after 26.777194 seconds
[apiclient] Waiting for at least one node to register and become ready
[apiclient] First node is ready after 0.503150 seconds
[apiclient] Creating a test deployment
[apiclient] Test deployment succeeded
[token-discovery] Created the kube-discovery deployment, waiting for it to become ready
[token-discovery] kube-discovery is ready after 3.002083 seconds
[addons] Created essential addon: kube-proxy
[addons] Created essential addon: kube-dns

Your Kubernetes master has initialized successfully!

But you still need to deploy a pod network to the cluster.
You should "kubectl apply -f" some pod network yaml file that's listed at:
    http://kubernetes.io/docs/admin/addons/

You can now join any number of machines by running the following on each node:

kubeadm join --token=e69295.cc499553b867df2c 192.168.255.6
```

And similar improvements to `kubeadm join` and `kubeadm reset`, and in different cases where things fail => still user-friendly output

As said earlier, the three last commits are the real ones. Please only look at them.
Here are the changes in a human-readable format:
 - Mark socat, ethtool and ebtables as soft deps, since kubelet can be run in a container.
 - Auto-start the kubelet service if it isn't active. This is really convenient. If kubeadm does it, it informs the user that it ran that command so the user knows what's happening.
 - Renamed /etc/kubernetes/cloud-config.json to /etc/kubernetes/cloud-config since it shouldn't be a json file
 - A lot of logging improvements
 - Removed dead code
 - Refactored the code so setting `KUBE_KUBERNETES_DIR` and `KUBE_HOST_PKI_PATH` actually works
 - Simplification of the code
 - Made a small logging/output framework:
   - `fmt.Println("[the-stage-here] Capital first letter of this message. Tell the user what the current state is")`
   - `fmt.Printf("[the-stage-here] Capital first letter. Maybe a [%v] in the end if an error should be displayed. Always ends with \n")`
   - `fmt.Errorf("Never starts with []. Includes a short error message plus the underlying error in [%v]. Never ends with \n")`
   - In short: made everything consistent, since now everything is done differently which is a mess...

File: cmd/kubeadm/app/preflight/checks.go [hunk 1]
```diff
@@ -346,3 +347,17 @@ func RunChecks(checks []PreFlightCheck, ww io.Writer) error {
...
+func TryStartKubelet() error {
+	// If we notice that the kubelet service is inactive, try to start it
+	initSystem, err := initsystem.GetInitSystem()
+	if err != nil {
+		fmt.Println("[preflight] No supported init system detected, won't check if kubelet is running")
+	} else if !initSystem.ServiceIsActive("kubelet") {
+		fmt.Printf("[preflight] Starting the kubelet service by running %q\n", "systemctl start kubelet")
+		if err := initSystem.ServiceStart("kubelet"); err != nil {
+			return fmt.Errorf("Couldn't start the kubelet service. Please start the kubelet service manually and try again.")
+		}
```
Comment (dgoodwin) [inline cmd/kubeadm/app/preflight/checks.go]: Will this error our if it's an systemd system, but kubelet is not managed by systemd? We should keep the door open for that possibility and let it work, so this should be a warning not an error. (I'm not sure if it will exit yet)

**Review by dgoodwin — COMMENTED** [review]
Just some quick notes but I can take another pass once this is rebased on the final version of the other three commits.

In trying to test it, it looks like my apiserver never comes up. I can't see where the problem is yet, there are no failed docker containers, it seems like it doesn't even start.

**Comment by luxas** [comment]
@dgoodwin Seems like there was a rebase conflict, but that's fixed now. PTAL

**Comment by dgoodwin** [comment]
@luxas looks healthy again here, I will give another pass once the other commits are in.

**Comment by luxas** [comment]
@mikedanese Please review this today and if possible add lgtm. See: https://github.com/kubernetes/kubernetes/pull/37568#issuecomment-265199717 for an explanation of what changes. This fixes a lot of bugs currently at HEAD as well.

@dgoodwin Ready for yet another pass.

When this is merged, which should be today or tomorrow, we can cut the next kubeadm release for v1.5 from current master.

File: cmd/kubeadm/app/cmd/init.go [hunk 2]
```diff
@@ -51,7 +52,11 @@ const (
 
 var (
 	initDoneMsgf = dedent.Dedent(`
-		Kubernetes master initialised successfully!
+		Your Kubernetes master has initialized successfully!
+
+		But you still need to deploy a pod network to the cluster.
+		You should "kubectl apply -f" some pod network yaml file that's listed at:
```
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/init.go]: Suggest:

You should now deploy a pod network to the cluster. 
Run "kubectl apply -f [podnetwork].yaml" with one of the options listed at:

File: cmd/kubeadm/app/preflight/checks.go [hunk 3]
```diff
@@ -357,3 +364,17 @@ func runChecks(checks []PreFlightCheck, ww io.Writer) error {
...
+func TryStartKubelet() {
+	// If we notice that the kubelet service is inactive, try to start it
+	initSystem, err := initsystem.GetInitSystem()
+	if err != nil {
+		fmt.Println("[preflight] No supported init system detected, won't check if kubelet is running")
+	} else if !initSystem.ServiceIsActive("kubelet") {
+
+		fmt.Printf("[preflight] Starting the kubelet systemd service by running %q\n", "systemctl start kubelet")
+		if err := initSystem.ServiceStart("kubelet"); err != nil {
+			fmt.Println("[preflight] Couldn't start the kubelet service via systemd. Please start the kubelet service manually and try again.")
```
Comment (dgoodwin) [inline cmd/kubeadm/app/preflight/checks.go]: Should we include the "err" here? Might be something relevant to the user in there.

Comment (dgoodwin) [inline cmd/kubeadm/app/preflight/checks.go]: Also not great that we're mentioning systemd multiple times here despite an attempt to remain init system agnostic. 

Why don't we just go to: 

Starting kubelet service...
WARNING: Unable to start kubelet service: %s
WARNING: Please ensure kubelet is running manually.

Drop "try again" as I think we just proceed here, and kubeadm will hang if it's not actually running.

File: cmd/kubeadm/app/cmd/init.go [hunk 4]
```diff
@@ -22,6 +22,7 @@ import (
 	"html/template"
 	"io"
 	"io/ioutil"
+	"os"
```
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/init.go]: With testing I notice that in this file we're outputting:

[preflight] Starting the kubelet systemd service by running "systemctl start kubelet"
Using Kubernetes version: v1.4.6

Where that is about the only line in all output that doesn't have a [something] prefix.

File: cmd/kubeadm/app/cmd/reset.go [hunk 5]
```diff
@@ -66,7 +66,7 @@ func NewReset(skipPreFlight, removeNode bool) (*Reset, error) {
 	if !skipPreFlight {
 		fmt.Println("[preflight] Running pre-flight checks...")
 
-		if err := preflight.RunResetCheck(); err != nil {
+		if err := preflight.RunChecks([]preflight.PreFlightCheck{preflight.IsRootCheck{}}, os.Stderr); err != nil {
```
Comment (dgoodwin) [inline cmd/kubeadm/app/cmd/reset.go]: Do you want to slip one more fix into this PR? :) 

We clean directories, then shutdown all running containers. This can result in etcd writing more to the directory after we cleaned it, and then your next init fails and you have to run reset a second time, which will work.

We need to cleanup directories after shutting down all containers.

I can get this in a followup PR if you prefer.

**Review by dgoodwin — COMMENTED** [review]
Couple more small changes, mostly text, testing looks good on my vms.

File: cmd/kubeadm/app/cmd/join.go [hunk 6]
```diff
@@ -109,25 +113,34 @@ func NewJoin(cfgPath string, args []string, cfg *kubeadmapi.NodeConfiguration, s
...
+		fmt.Println("[preflight] Running pre-flight checks...")
+
+		// First, check if we're root separately from the other preflight checks and fail fast
+		if err := preflight.RunChecks([]preflight.PreFlightCheck{preflight.IsRootCheck{}}, os.Stderr); err != nil {
+			return nil, &preflight.PreFlightError{Msg: err.Error()}
+		}
+
+		// Then continue with the others...
+		if err := preflight.RunJoinNodeChecks(cfg); err != nil {
 			return nil, &preflight.PreFlightError{Msg: err.Error()}
```
Comment (mikedanese) [inline cmd/kubeadm/app/cmd/join.go]: why don't we just return a preflight.PreFlightError from RunJoinNodeChecks.

File: cmd/kubeadm/app/cmd/join.go [hunk 7]
```diff
@@ -109,25 +113,34 @@ func NewJoin(cfgPath string, args []string, cfg *kubeadmapi.NodeConfiguration, s
...
 	}
 
 	if !skipPreFlight {
-		fmt.Println("Running pre-flight checks")
-		err := preflight.RunJoinNodeChecks(cfg)
-		if err != nil {
+		fmt.Println("[preflight] Running pre-flight checks...")
+
+		// First, check if we're root separately from the other preflight checks and fail fast
+		if err := preflight.RunChecks([]preflight.PreFlightCheck{preflight.IsRootCheck{}}, os.Stderr); err != nil {
```
Comment (mikedanese) [inline cmd/kubeadm/app/cmd/join.go]: Consider moving into preflight package.

File: cmd/kubeadm/app/master/apiclient.go [hunk 8]
```diff
@@ -271,9 +269,11 @@ func createDummyDeployment(client *clientset.Clientset) {
 		return true, nil
 	})
 
-	fmt.Println("<master/apiclient> test deployment succeeded")
+	fmt.Println("[apiclient] Test deployment succeeded")
 
-	if err := client.Extensions().Deployments(api.NamespaceSystem).Delete("dummy", &v1.DeleteOptions{}); err != nil {
```
Comment (mikedanese) [inline cmd/kubeadm/app/master/apiclient.go]: Let's just not create if we are not going to delete it.

Comment (mikedanese) [inline cmd/kubeadm/app/master/apiclient.go]: Have you checked if https://github.com/kubernetes/kubernetes/pull/38330 fixes your problem? Can we just put a priority on getting that merged?

File: cmd/kubeadm/app/cmd/init.go (hunk 4 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: That was a rebase conflict, thanks for catching

File: cmd/kubeadm/app/cmd/init.go (hunk 2 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/init.go]: Fixed, thanks

File: cmd/kubeadm/app/master/apiclient.go (hunk 8 above)
Comment (luxas) [inline cmd/kubeadm/app/master/apiclient.go]: It does, tested it now.
Gonna revert this and wait for #38330, but hopefully it will merge very soon

File: cmd/kubeadm/app/cmd/join.go (hunk 7 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/join.go]: Yes, gonna do that

File: cmd/kubeadm/app/cmd/join.go (hunk 6 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/join.go]: Seems like no one just saw it before, fixing

File: cmd/kubeadm/app/cmd/reset.go (hunk 5 above)
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: I'll fix it

**Comment by luxas** [comment]
Thanks for the reviews @mikedanese and @dgoodwin

If there's anything else you think should be fixed, please comment now. Otherwise I'll rebase and make it ready for merge tomorrow so we can get it in in time.

**Comment by luxas** [comment]
@dgoodwin Looks ok?

**Comment by mikedanese** [comment]
Please squash fixup commits

**Comment by luxas** [comment]
@mikedanese The fourth and the fifth one? Sure.
Also, seems like I have to update bazel as well...
